rfcs/git-sdk-requirements/RFC.md

```diff
@@ -0,0 +1,697 @@
+pull() // ✅ https://stackoverflow.com/questions/39651287/doing-a-git-pull-with-libgit2
+sparseCheckout() // ❓ https://github.com/libgit2/libgit2/issues/2263 (open pr, need to build, test)
+
+// branch related
+currentBranch() // ✅ https://stackoverflow.com/questions/12132862/how-do-i-get-the-name-of-the-current-branch-in-libgit2
+createBranch() // ✅ https://libgit2.org/libgit2/#HEAD/group/branch/git_branch_create
+renameBranch() // ✅ https://libgit2.org/libgit2/#v0.17.0/group/branch
+switchBranch() // ✅ https://stackoverflow.com/questions/46757991/checkout-branch-with-libgit2
+deleteBranch() // ✅ https://libgit2.org/libgit2/#v0.17.0/group/branch/git_branch_delete
+
+// change related
+// (3 API different "changes" concepts...)
+unstagedChanges() // ❓ didn't find example with this, but it may exist
+uncommittedChanges() // ❓ didn't find example too
+unpushedChanges() // ✅ https://stackoverflow.com/questions/42131934/get-unpushed-commits-with-libgit2
+
+// host like GitHub or GitLab dependent
+// ouside of libgit2 scope
+// would need to be built as part of Git SDK and send relevant commands to libgit2
+signIn()
+signOut()
+createFork()
+syncFork()
+openPullRequest()
+```
+
+#### Q: Is 833 KB WASM file an issue to load?
+
+Should be non issue as its under 1 MB.
+
+#### Q: Can you keep the file system inside WASM too?
+
+No, you can't. There is no file system in the WebAssembly runtime.
+
+#### Q: How does the architecture with this approach look?
+
+[WASM Git](https://github.com/petersalomonsen/wasm-git) reccomends using it through a [web worker](https://github.com/petersalomonsen/wasm-git#example-webworker-with-pre-built-binaries). For unclear benefit.
+
+> note: need to understand the reasoning more, perhaps there is a good reason.
+
+When you compile libgit2 to wasm with Emscripten, as build artefacts, you get `lg2.wasm` file that is the libgit2 itself. And also `lg2.js`.
+
+The purpose of `lg2.js` file is to provide the file system for libgit2 to sync with.
+
+The code for `lg2.js` is quite hard to read and would need to be rewritten to be more nice to use and iterate on.
+
+It does already do these 2 things that are useful:
+
+1. Exposes a `libgit` variable with a `callMain` function.
+
+You would then do:
+
+```js
+FS.mkdir("/")
+FS.mount(MEMFS, {}, "/")
+FS.chdir("/")
+const result = libgit.callMain(["clone", "https://github.com/inlang/inlang.git", "inlang"])
+FS.chdir("inlang")
+```
+
+You can expose different C functions that run the commands. For example, we have built [open pr for shallow clones](https://github.com/libgit2/libgit2/pull/6396) and exposed it via a new method for `callMain`:
+
+```js
+const result = libgit.callMain(["shallowClone", "https://github.com/inlang/inlang.git", "inlang"])
+```
+
+That function was same as clone, just also set `clone_opts.fetch_opts.depth` option before doing the clone. Extending libgit2 with new commands would work in the same way.
```

Comment:
This is a usage explanation. Not an architectural exploration. 

What happens under the hood? What approach will make the usage of the git-sdk easier? The surface APIs like `libgit.callMain` can always be changed.